### ISSUE TITLE
stack overflow during type checking of badly typed function

### ISSUE THREAD

**Issue by riffraff** [elm-lang/elm-compiler#349]
I am new to elm and don't know haskell either, so I accidentally typed this:

    iterate n f x =
        if n == 0 then x
        else iterate(n - 1, f ,f(x))

    square x = x * x 

    main = asText (iterate 1 square 3)

notice the error in line 3. 

This causes the compiler to fail like this:

     $ elm ibroken.eml
     [1 of 1] Compiling Main                ( ibroken.eml )
     Stack space overflow: current size 8388608 bytes.
     Use `+RTS -Ksize -RTS' to increase it.

As a side note, this affects the online interpreter/repl, since the client never receives a response from the server.

**Comment by evancz** [comment]
In the latest version of Elm, this code is triggering the following error.

```
-- TYPE ERROR --------------------------------------------------------- hats.elm

This expression is triggering a type mismatch.

1| 
2|     if n == 0
3|       then x
4|       else iterate(n - 1, f ,f(x))
5| 

To be more specific, as I infer all the types, I am seeing a conflict between
this type:

    (number, a -> b, b)

and this type:

    number

```

So I am going to mark this as resolved. I think we'll be able to give hints along the lines of "argument number 1 is messed up" with a project coming up this summer. cc @JoeyEremondi and adding this example to https://github.com/evancz/error-message-catalog as common starter error that it's important to handle well!
